Newsletters: add 280 (2023-12-06)

harding opened the pull request:
- [x] Lede, releases/RCs, topic entries @harding 
- [x] Bitcoin Core PR Review Club @LarryRuane 
- [x] Bitcoin Core 28848 @glozow 
- [x] LDK 2540 @bitschmidty

vostrnad commented on _posts/en/newsletters/2023-12-06-newsletter.md:
```diff
@@ -0,0 +1,179 @@
+release candidates.*
+
+- [Bitcoin Core 26.0rc3][] is a release candidate for the next major
+  version of the predominant full-node implementation. There's a [testing
```
Per https://github.com/bitcoinops/bitcoinops.github.io/pull/1400#discussion_r1402368137:
```suggestion
  version of the predominant full node implementation. There's a [testing
```

LarryRuane replied:
It reads better to me with the hyphen, because it's conceptually a single, unitary thing (rather than saying it's a node that happens to be "full" whatever that means -- is it out of disk space?), but I don't have a strong opinion.

vostrnad replied:
I'm not knowledgeble enough to discuss the language perspective, I'm just saying this goes against long established terminology both at Bitcoin Optech and most everywhere else I could find. The hyphenated version was never used on Optech until a few weeks ago. I'd think we'd like to keep terminology consistent since we even have a style guide for this very purpose.

harding replied:
@LarryRuane would you hyphenate "Bitcoin node implementation" and "LN node implementation"?

Given that Optech frequently discusses things like peer networks that can be thought of topologically as represented by nodes and edges, we modify the noun "node" a lot to make it clear what type of node we're talking about, so we probably want a general policy.

murchandamus commented on _posts/en/newsletters/2023-12-06-newsletter.md:
```diff
@@ -0,0 +1,177 @@
+  chunks each about 4 million weight in size, each of which could be
+  mined successively into a block.
+
+    By comparing different lists of chunks, we can determine which would
```
"Chunkings across multiple clusters" would only occur when we are trying to determine whether a replacement is strictly better than our current mempool content. E.g. a replacement transaction could combine two previously unconnected clusters, and then we’d have to create combined feerate diagram across both these clusters to compare the state with the two separate clusters to the state after the replacement.

murchandamus commented on _posts/en/newsletters/2023-12-06-newsletter.md:
```diff
@@ -0,0 +1,177 @@
+  _Clusters_ of one or more related transactions are designed to be easy
+  to include into feerate-sorted _chunks_ that are each independent of
+  the other chunks. For example, we might split a mempool into 100
+  chunks each about 4 million weight in size, each of which could be
```
Having calculated a linearization for every cluster in our mempool and subdivided it into chunks, we can now create a total order of all chunks in our mempool. We only need to compare the best unselected chunks in each cluster, rather than each transaction with its ancestry. By picking the best unselected chunk from a cluster repeatedly, we can easily assemble a block template, and vice versa, if we evict, we can simply compare the worst feerate chunks from each cluster to see which one we would pick into a block last.

ajtowns replied:
@murchandamus Note that to assemble the block we're pick the best chunks from *all* clusters in the mempool, simply comparing them by feerate (and perhaps size).

@sipa It makes sense to me that chunks need to be limited in size (as we add chunks to a block as individual units), but not why clusters necessarily need to have the same limit. I think the cluster limit results from the possibility that the optimal chunking of a cluster may end up being almost the entire cluster (particularly when multiple children pay for a parent, eg), and if we didn't limit the cluster size, we could end up stuck with a suboptimal linearisation, which we don't want. Maybe some commentary about limits would make for a good post as well?

murchandamus commented on _posts/en/newsletters/2023-12-06-newsletter.md:
```diff
@@ -0,0 +1,177 @@
+      chunking must exist which is strictly better than both".  That
+      means an effective method for merging two different incomparable
+      chunkings can be a powerful tool for improving miner
+      profitability.  For example, miners Alice and Bob could both try
```
Not just miners, but any node could try to find better linearizations

murchandamus commented on _posts/en/newsletters/2023-12-06-newsletter.md:
```diff
@@ -0,0 +1,177 @@
+  ordering of Bitcoin transactions requires parent transactions get
+  confirmed before their children, either by putting the parent in an
+  earlier block than any of its children, or earlier in the same block.
+  _Clusters_ of one or more related transactions are designed to be easy
+  to include into feerate-sorted _chunks_ that are each independent of
+  the other chunks. For example, we might split a mempool into 100
```
I think there might be misunderstanding here. A _cluster_ refers to the connected component of a transaction, i.e. all ancestors and descendants of a transaction all of the ancestors and descendants of all those, plus all of the ancestors and descendants of all that, etc. until there is nothing to add. _Chunks_ refer to sections of a linearization of a cluster. 

Linearizations must be topologically valid. They are chunked by first making every transaction its own chunk, then combining any two chunks A and B, if A is the left neighbor of B and B has a higher feerate than A. Ideally, linearizations are sorted to produce the highest possible feerate diagram across the cluster, but non-optimal linearizations may occur.

We are currently expecting that clusters might be limited to 100 transactions, much like the descendant and ancestor limit in the current mempool design, and since chunks are subcomponents of a cluster’s linearization, they would usually not be whole blocks.

murchandamus commented on _posts/en/newsletters/2023-12-06-newsletter.md:
```diff
@@ -0,0 +1,179 @@
+  Delving Bitcoin with the results of some simulations he's run using
+  [warnet][], a program that launches a large number of Bitcoin nodes
+  with a defined set of connections between them (usually on a test
+  network).  Zipkin results show the extra memory that would be used if
```
```suggestion
  network).  Zipkin's results show the extra memory that would be used if
```

murchandamus commented on _posts/en/newsletters/2023-12-06-newsletter.md:
```diff
@@ -0,0 +1,179 @@
+projects.  Please consider upgrading to new releases or helping to test
+release candidates.*
+
+- [Bitcoin Core 26.0rc3][] is a release candidate for the next major
```
rc4 and final got tagged meanwhile

instagibbs commented on _posts/en/newsletters/2023-12-06-newsletter.md:
```diff
@@ -0,0 +1,179 @@
+      third chunking that was at least as good as either of their
+      original chunkings---and possibly better.
+
+    - [Post-cluster package RBF][cluster rbf] discusses an alternative
```
you might want to note we don't even *have* package rbf today, and what is proposed pre-cluster mempool is very limited

harding commented:
Thanks @murchandamus @sipa @instagibbs for the cluster mempool comments so far!  I'm starting on edits now and should get them pushed in a few hours.

LarryRuane commented on _posts/en/newsletters/2023-12-06-newsletter.md:
```diff
@@ -0,0 +1,248 @@
+  group][wg-cluster-mempool] (WG) on the Delving Bitcoin forum. Cluster
+  mempool is a proposal to make it much easier to compare the
+  profitability of different valid orderings of transactions.  A valid
+  ordering of Bitcoin transactions requires parent transactions get
```
```suggestion
  ordering of Bitcoin transactions requires parent transactions to be
```

LarryRuane commented on _posts/en/newsletters/2023-12-06-newsletter.md:
```diff
@@ -0,0 +1,248 @@
+    By comparing different lists of chunks, we can determine which would
+    be better for miners.  Any change that increases the total fees of
+    the first chunk (in the same number of vbytes or less) is an
+    improvement to a miner; any change that does the opposite (reduce
```
```suggestion
    improvement to a miner; any change that does the opposite (reduces
```

LarryRuane commented on _posts/en/newsletters/2023-12-06-newsletter.md:
```diff
@@ -0,0 +1,248 @@
+      means an effective method for merging two different incomparable
+      chunkings can be a powerful tool for improving miner
+      profitability.  For example, miners Alice and Bob could both try
+      to find the optimal chunking and could share their results with
```
```suggestion
      to find the optimal chunking and share their results with
```

LarryRuane commented on _posts/en/newsletters/2023-12-06-newsletter.md:
```diff
@@ -0,0 +1,248 @@
+
+- [Bitcoin Core #28848][] updates the `submitpackage` RPC to be more helpful when
+  any transaction fails. Instead of throwing a JSONRPCError with the first
+  failure, it return results for each transaction whenever possible.
```
```suggestion
  failure, it returns results for each transaction whenever possible.
```

harding commented:
@sipa @murchandamus @instagibbs Thanks again!  The following two commits make edits based on the cluster mempool feedback (no other feedback is addressed in these commits):

- https://github.com/bitcoinops/bitcoinops.github.io/pull/1421/commits/48c66753f4c302393ef9491a428022288bafbf21 is edits to the news item
- https://github.com/bitcoinops/bitcoinops.github.io/pull/1421/commits/17438717e3739cb7a1788ec8437b55b050c1b612 updates the topic page for _cluster mempool_ to use an edited version of @sipa's excellent summary from this thread

@sipa

One question I had about your summary was this sentence: "However, it is hopefully apparent that the limit of how much can be affected by a transaction is just whatever transactions it is directly or indirectly connected to - clusters."  I changed that to "[...] connected to, i.e. its cluster".  Is that edit consistent with your intention?

Everyone else: sorry I haven't had time to make edits for your feedback yet.  I'll get to it Tuesday (my time).

murchandamus commented on _topics/en/cluster-mempool.md:
```diff
@@ -52,20 +52,98 @@ see_also:
+
+However, it is hopefully apparent that the limit of how much can be
+affected by a transaction is just whatever transactions it is directly
+or indirectly connected to, i.e. its cluster.<!--FIXME:harding confirm edit--> Optimal chunks never cross
+cluster boundaries---any chunk that did could be split into independent
```
I feel this may be a bit unclear at this stage of the explanation, but just as an overall "encyclopedic definition". A chunk is a subset of cluster, therefore must always consist of transactions that are related in some sense that makes them part of the same cluster, but within an optimal linearization, each chunk ends up necessarily consisting of transactions connected within the chunk.

bitschmidty commented on _posts/en/newsletters/2023-12-06-newsletter.md:
```diff
@@ -0,0 +1,256 @@
+answers.  Click on a question below to see a summary of the answer from
+the meeting.*
+
+Before each [major Bitcoin Core release][], extensive testing by the
```
I think the summary for this PR Review Club is great and we could likely drop all of the Q&A section below, which I thought added less value.

LarryRuane replied:
> I think the summary for this PR Review Club is great and we could likely drop all of the Q&A section below, which I thought added less value.

Thanks, and I'm perfectly okay with that suggestion. In fact, I almost did that, but thought that _maybe_ keeping it would give the reader a feel for the level and kinds of detail and depth that occur in these release testing meetings -- I agree that the actual content is pretty low-information. But OTOH the reader can simply click the link to view the actual discussion log, so maybe that's not a good reason.

I just noticed, that there is no actual link to the review club page, I'll fix that right now (with a non-force push, new commit).

Feel free to remove the questions, or I can do that, or we can wait to see what others think. But I do like your suggestion.

harding replied:
I'm +0.1 on keeping the QA.  I'm happy with whatever Larry and Mike decide.

harding commented on _topics/en/cluster-mempool.md:
```diff
@@ -52,20 +52,98 @@ see_also:
+and ideally, with ~millisecond latency to not stall other processing).
+
+An obvious question arises: can't we precompute things so that computing
+the updated mining-order-of-full-mempool isn't too slow? And we kind of
```
Leaving as-is.  It's true that my professional writing almost never includes hypophora, but we've used it in other contributed documentation (and the topic pages were always meant to include community contributions).  If someone dislikes it enough, they can edit the document themselves.  :-)

harding commented on _topics/en/cluster-mempool.md:
```diff
@@ -52,20 +52,98 @@ see_also:
+affected by a transaction is just whatever transactions it is directly
+or indirectly connected to, i.e. its cluster.<!--FIXME:harding confirm edit--> Optimal chunks never cross
+cluster boundaries---any chunk that did could be split into independent
+chunks, and doing so would never worsen the result.
```
Updated the intro paragraph to the text below:

>   **Cluster mempool** is a proposal to associate each unconfirmed
  transaction in a mempool with related transactions, creating a
  _cluster_.  Each cluster contains feerate-sorted _chunks_ consisting
  of one or more transactions.  Thus, if clusters are limited in size,
  we also limit how much needs to be recomputed in response to new
  transactions being added to the mempool, allowing algorithmic
  decisions affecting the entire mempool to to complete fast enough to
  use them in P2P network code.

(The other revisions were required because we limit the first-paragraph to 500 characters in order to be able to use them in the future in something like a glossary.)

harding commented:
Made edits for all feedback, added lede, updated releases/RCs.  I quickly reviewed the PR Review Club summary (thanks @LarryRuane).  I haven't yet added topic entries.  I also punted all unwritten PR merge summaries to next week's newsletter.

vostrnad commented on _posts/en/newsletters/2023-12-06-newsletter.md:
```diff
@@ -0,0 +1,283 @@
+
+{% include functions/details-list.md
+  q0="Which operating systems are people running?"
+  a0="Ubuntu 22.04 on WSL (Windows subsystem for Linux); macOs 13.4 (m1 chip)."
```
```suggestion
  a0="Ubuntu 22.04 on WSL (Windows Subsystem for Linux); macOS 13.4 (M1 chip)."
```

vostrnad commented on _posts/en/newsletters/2023-12-06-newsletter.md:
```diff
@@ -0,0 +1,283 @@
+projects.  Please consider upgrading to new releases or helping to test
+release candidates.*
+
+- [Bitcoin Core 26.0][] is a the next major version of the predominant
+  full-node implementation.  The release includes experimental support
```
```suggestion
- [Bitcoin Core 26.0][] is the next major version of the predominant
  full node implementation.  The release includes experimental support
```

vostrnad commented on _posts/en/newsletters/2023-12-06-newsletter.md:
```diff
@@ -0,0 +1,283 @@
+[Bitcoin Inquisition][bitcoin inquisition repo].*
+
+- [Bitcoin Core #28848][] updates the `submitpackage` RPC to be more helpful when
+  any transaction fails. Instead of throwing a JSONRPCError with the first
```
```suggestion
  any transaction fails. Instead of throwing a `JSONRPCError` with the first
```

vostrnad commented on _topics/en/cluster-mempool.md:
```diff
@@ -52,20 +52,102 @@ see_also:
-hoped that this will allow transaction relay to provide more flexible
-policy without exposing relay nodes to new resource-wasting attacks.
+The overall goal of cluster mempool is being able to reason about the
+effect of transactions on the blocks a miner would create if it an
```
```suggestion
effect of transactions on the blocks a miner would create if it has an
```

vostrnad commented on _topics/en/cluster-mempool.md:
```diff
@@ -52,20 +52,102 @@ see_also:
+An obvious question arises: can't we precompute things so that computing
+the updated selection-order-of-full-mempool isn't too slow? And we kind of
+can. What the block building algorithm ultimately does is find groups of
+transactions to include at once (e.g. child pays for parent means both
```
```suggestion
transactions to include at once (e.g. [child pays for parent][topic cpfp] means both
```

LarryRuane commented on _topics/en/cluster-mempool.md:
```diff
@@ -52,20 +55,102 @@ see_also:
-package relay] or [replacement][topic rbf] should be accepted.  It is
-hoped that this will allow transaction relay to provide more flexible
-policy without exposing relay nodes to new resource-wasting attacks.
+The overall goal of cluster mempool is being able to reason about the
```
```suggestion
The overall goal of cluster mempool is to be able to reason about the
```

LarryRuane commented on _topics/en/cluster-mempool.md:
```diff
@@ -52,20 +55,102 @@ see_also:
+effect of transactions on the blocks a miner would create if it has an
+identical mempool to the local node's mempool.
+
+The most apparent example for why we have a need for that kind of
```
```suggestion
The most apparent example of why we need that kind of
```

LarryRuane commented on _topics/en/cluster-mempool.md:
```diff
@@ -52,20 +55,102 @@ see_also:
+
+There is an obvious solution to the problem: instead of finding
+lowest-feerate-descendant-sets, run the normal selection algorithm on the
+entire mempool (don't stop after one block worth of transactions), and
```
```suggestion
entire mempool (don't stop after one block's worth of transactions), and
```

LarryRuane commented on _topics/en/cluster-mempool.md:
```diff
@@ -52,20 +55,102 @@ see_also:
+see what it would include last. That's the thing you should evict!
+
+Sadly, this is computationally infeasible. The block building algorithm
+is fairly fast when running on one block worth of transactions, but
```
```suggestion
is fairly fast when running on one block's worth of transactions, but
```

LarryRuane commented on _topics/en/cluster-mempool.md:
```diff
@@ -52,20 +55,102 @@ see_also:
+entire mempool (don't stop after one block worth of transactions), and
+see what it would include last. That's the thing you should evict!
+
+Sadly, this is computationally infeasible. The block building algorithm
```
```suggestion
Sadly, this is computationally infeasible. The block-building algorithm
```
(and 4 more the same below)

LarryRuane commented on _topics/en/cluster-mempool.md:
```diff
@@ -52,20 +55,102 @@ see_also:
+Sadly, this is computationally infeasible. The block building algorithm
+is fairly fast when running on one block worth of transactions, but
+running it on the entire mempool would take a significant multiple of
+the time. This is absolutely not possible to do every time
```
```suggestion
the time. This is not possible to do every time
```

LarryRuane commented on _topics/en/cluster-mempool.md:
```diff
@@ -52,20 +55,102 @@ see_also:
+will be included at once), at the feerate that included set overall has
+(sum of fees divided by sum of sizes). If we could somehow precompute
+those groupings (which we'll call chunks) ahead of time, then the actual
+block building algorithm (ignoring binpacking issues once we get close
```
```suggestion
block building algorithm (ignoring bin-packing issues once we get close
```

LarryRuane commented on _topics/en/cluster-mempool.md:
```diff
@@ -52,20 +55,102 @@ see_also:
-  or removing new clusters to complete fast enough to use them in P2P
-  network code.
+  _cluster_.  Each cluster contains feerate-sorted _chunks_ consisting
+  of one or more transactions.  If clusters are limited in size,
```
```suggestion
  of one or more transactions.  If we limit a cluster's size,
```
(otherwise the size sounds exgengous, which I don't think is the intention?)

LarryRuane commented on _topics/en/cluster-mempool.md:
```diff
@@ -52,20 +55,102 @@ see_also:
+mempool exceeds the node's size limit) can choose to evict the very best
+transaction in the mempool overall ([example][wuille slide7]).
+
+To describe the current limitation in short: miners prefer to select the
```
```suggestion
To describe the current limitation in short: miners prefer to select for inclusion the
```
(otherwise, at first I read this as selecting for eviction, since that was just being mentioned.)

LarryRuane commented on _topics/en/cluster-mempool.md:
```diff
@@ -52,20 +55,102 @@ see_also:
+the updated selection-order-of-full-mempool isn't too slow? And we kind of
+can. What the block building algorithm ultimately does is find groups of
+transactions to include at once (e.g. [child pays for parent][topic cpfp] means both
+will be included at once), at the feerate that included set overall has
```
```suggestion
will be included at once), at the feerate that the included set overall has
```

LarryRuane commented on _topics/en/cluster-mempool.md:
```diff
@@ -52,20 +55,102 @@ see_also:
+
+So can we precompute the chunks, and update just the ones that get
+affected when a new transaction or block comes in? Sadly, there is in
+theory no bound on how many transactions' chunkings can get affected by
```
```suggestion
theory no bound on how many transactions' chunkings can be affected by
```

LarryRuane commented on _posts/en/newsletters/2023-12-06-newsletter.md:
```diff
@@ -0,0 +1,282 @@
+  group][wg-cluster-mempool] (WG) on the Delving Bitcoin forum. Cluster
+  mempool is a proposal to make it much easier to operate on the mempool
+  while respecting the required ordering of transactions.  A valid
+  ordering of Bitcoin transactions requires parent transactions be
```
```suggestion
  ordering of Bitcoin transactions requires parent transactions to be
```

LarryRuane commented on _posts/en/newsletters/2023-12-06-newsletter.md:
```diff
@@ -0,0 +1,282 @@
+  confirmed before their children, either by putting the parent in an
+  earlier block than any of its children, or earlier in the same block.
+  In the cluster mempool design, _clusters_ of one or more related
+  transactions are designed to be to split into feerate-sorted _chunks_.
```
```suggestion
  transactions are designed to be split into feerate-sorted _chunks_.
```

LarryRuane commented on _posts/en/newsletters/2023-12-06-newsletter.md:
```diff
@@ -0,0 +1,282 @@
+         inserting the new transaction in a location that is valid
+         (parents before children).  This has the advantage that it
+         preserves any existing optimizations in the unmodified chunks
+         but the downside that it could place the transaction in a
```
```suggestion
         but the downside is that it could place the transaction in a
```

LarryRuane commented on _posts/en/newsletters/2023-12-06-newsletter.md:
```diff
@@ -0,0 +1,282 @@
+         suboptimal location.
+
+      After the two different types of updates have been made, a
+      comparison may reveal one of them is strictly better, in which
```
```suggestion
      comparison may reveal one of them to be strictly better, in which
```
or maybe "... may reveal that one of them is strictly better..."

LarryRuane commented on _posts/en/newsletters/2023-12-06-newsletter.md:
```diff
@@ -0,0 +1,282 @@
+  network).  Zipkin's results show the extra memory that would be used if
+  several proposed changes to peer management code are merged (either
+  independently or together).  He also notes that he's excited to use
+  simulations for testing other proposed changes and to quantify the
```
```suggestion
  simulations for testing other proposed changes and quantifying the
```

bitschmidty commented:
Took some of @LarryRuane's and @vostrnad's suggestions, left others.

Resolved many of the feedback/suggestion comments in the PR, but left other threads open for better discoverability of interesting content from this PR.

Squashed commits and merged. Thank you @harding @LarryRuane @glozow @sipa @ajtowns @vostrnad @murchandamus @instagibbs for your contributions this week! 🚀
